sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -163,6 +163,155 @@ class CollationStringExpressionsSuite
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci").flatMap(
+      c => Seq(
+        SubstringTestCase("select left(' a世a ' collate " + c + ", '3')", c, Row(" a世")), // scalastyle:ignore
+        SubstringTestCase("select right(' a世a ' collate " + c + ", '3')", c, Row("世a ")), // scalastyle:ignore
+        SubstringTestCase("select right('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null)", c, Row(null)),
+        SubstringTestCase("select substr('' collate " + c + ", null, null)", c, Row(null)),
+        SubstringTestCase("select left('' collate " + c + ", null)", c, Row(null))
+      )
+    )
+    checks.foreach(check => {
+      // Result & data type
+      checkAnswer(sql(check.query), check.result)
+      assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+    })
+  }
+
+  test("Support Left/Right/Substr with session collation") {
+    case class SubstringTestCase(query: String, collation: String, result: Row)
+    val checks = Seq("utf8_binary_lcase", "utf8_binary", "unicode", "unicode_ci")
+      .flatMap { c =>
+        Seq(
+          SubstringTestCase("select left('abc', 1)", c, Row("a")),
+          SubstringTestCase("select right('def', 1)", c, Row("f")),
+          SubstringTestCase("select substr('ghi', 1)", c, Row("ghi"))
+        )
+      }
+    checks.foreach(check => {
+      withSQLConf(SqlApiConf.DEFAULT_COLLATION -> check.collation) {
+        // Result & data type
+        checkAnswer(sql(check.query), check.result)
+        assert(sql(check.query).schema.fields.head.dataType.sameType(StringType(check.collation)))
+      }
+    })
+  }
+
+  test("Support Left/Right/Substr on struct fields with collation") {
+    Seq(None, Some("utf8_binary_lcase"), Some("utf8_binary"), Some("unicode"), Some("unicode_ci"))
+      .zipWithIndex
+      .foreach {
+        case (collationNameMaybe, i) =>
+          withTable(s"t1234$i") {
+            sql(s"CREATE TABLE t1234$i(i STRING, s" +
+              " struct<a: string" +
+              collationNameMaybe.map(cn => " collate " + cn).getOrElse("") +
+              ">) USING parquet")
+            (1 to 5).map(n => "a" + " " * n).foreach { v =>
+              sql(s"INSERT OVERWRITE t1234$i VALUES ('1', named_struct('a', '$v'))")
+            }
+            assert(sql(s"SELECT i, left(s.a, 1) FROM t1234$i").schema(1).dataType ==
+              collationNameMaybe.map(cn =>
+                StringType(cn)).getOrElse(StringType))
+            assert(sql(s"SELECT i, right(s.a, 1) FROM t1234$i").schema(1).dataType ==
+              collationNameMaybe.map(cn =>
+                StringType(cn)).getOrElse(StringType))
+            assert(sql(s"SELECT i, substr(s.a, 1, 0) FROM t1234$i").schema(1).dataType ==
+              collationNameMaybe.map(cn =>
+                StringType(cn)).getOrElse(StringType))
+          }
+      }
+  }
```

Comment:
why do we need this test case?

Reply:
<img width="491" alt="image" src="https://github.com/apache/spark/assets/31429832/57ca4f80-0b5f-438a-9d01-bc4dbf3e0834">

Reply:
Above is the reason. I removed the test, but if you want me to reinstate it just let me know.